packages/typir/src/graph/type-node.ts

```diff
@@ -55,6 +81,225 @@ export abstract class Type {
     abstract getUserRepresentation(): string;
 
 
+
+    // store the state of the initialization process of this type
+
+    protected initializationState: TypeInitializationState = 'Invalid';
+
+    getInitializationState(): TypeInitializationState {
+        return this.initializationState;
+    }
+
+    protected assertState(expectedState: TypeInitializationState): void {
+        if (this.isInState(expectedState) === false) {
+            throw new Error(`The current state of type '${this.identifier}' is ${this.initializationState}, but ${expectedState} is expected.`);
+        }
+    }
+    protected assertNotState(expectedState: TypeInitializationState): void {
+        if (this.isNotInState(expectedState) === false) {
+            throw new Error(`The current state of type '${this.identifier}' is ${this.initializationState}, but this state is not expected.`);
+        }
+    }
+    protected assertStateOrLater(expectedState: TypeInitializationState): void {
+        if (this.isInStateOrLater(expectedState) === false) {
+            throw new Error(`The current state of type '${this.identifier}' is ${this.initializationState}, but this state is not expected.`);
+        }
+    }
+
+    isInState(state: TypeInitializationState): boolean {
+        return this.initializationState === state;
+    }
+    isNotInState(state: TypeInitializationState): boolean {
+        return this.initializationState !== state;
+    }
+    isInStateOrLater(state: TypeInitializationState): boolean {
+        switch (state) {
+            case 'Invalid':
+                return true;
+            case 'Identifiable':
+                return this.initializationState !== 'Invalid';
+            case 'Completed':
+                return this.initializationState === 'Completed';
+            default:
+                assertUnreachable(state);
+        }
+    }
+
+
+    // manage listeners for updates of the initialization state
+
+    protected stateListeners: TypeStateListener[] = [];
+
+    addListener(listener: TypeStateListener, informIfNotInvalidAnymore: boolean): void {
+        this.stateListeners.push(listener);
+        if (informIfNotInvalidAnymore) {
```

Comment:
You are right in general. Here I don't see the purpose, since the notifications are always done in a different way, i.e. there is no reuse possible. Or do I miss something?